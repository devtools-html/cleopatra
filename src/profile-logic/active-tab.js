/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { getThreadSelectors } from '../selectors/per-thread';
import { assertExhaustiveCheck } from '../utils/flow';

import type { State } from '../types/state';
import type { ThreadIndex, Pid } from '../types/profile';
import type {
  GlobalTrack,
  LocalTrack,
  TrackIndex,
} from '../types/profile-derived';

/**
 * Take the global tracks and decide which one to hide during the active tab view.
 * Some global tracks are allowed, some tracks are not, and we have to do some
 * computations for some('process' type specifically).
 */
export function computeActiveTabHiddenGlobalTracks(
  globalTracks: GlobalTrack[],
  state: State
): Set<TrackIndex> {
  const activeTabHiddenGlobalTracks = new Set();

  for (let trackIndex = 0; trackIndex < globalTracks.length; trackIndex++) {
    const globalTrack: GlobalTrack = globalTracks[trackIndex];
    const trackType = globalTrack.type;

    switch (trackType) {
      case 'screenshots':
        // Do not hide screenshots.
        break;
      case 'visual-progress':
      case 'perceptual-visual-progress':
      case 'contentful-visual-progress':
        // Hide those global track types because we want to hide as much as
        // possible from web developers for now.
        activeTabHiddenGlobalTracks.add(trackIndex);
        break;
      case 'process': {
        // Do not display empty tracks if the tab filtered thread is empty.
        if (
          globalTrack.mainThreadIndex !== undefined &&
          globalTrack.mainThreadIndex !== null &&
          isTabFilteredThreadEmpty(globalTrack.mainThreadIndex, state)
        ) {
          // Thread is empty and we should hide it.
          activeTabHiddenGlobalTracks.add(trackIndex);
        }
        break;
      }
      default:
        throw assertExhaustiveCheck(trackType, `Unhandled GlobalTrack type.`);
    }
  }

  return activeTabHiddenGlobalTracks;
}

/**
 * Take the local tracks and decide which one to hide during the active tab view.
 * Some tracks are not allowed, and we have to do some computations for some('thread' type specifically).
 */
export function computeActiveTabHiddenLocalTracksByPid(
  localTracksByPid: Map<Pid, LocalTrack[]>,
  state: State
): Map<Pid, Set<TrackIndex>> {
  const activeTabHiddenLocalTracksByPid = new Map();

  for (const [pid, localTracks] of localTracksByPid) {
    // Pre-put the new Set here, because we should keep the whole processes even though they are empty.
    const currentLocalTracks = new Set();
    activeTabHiddenLocalTracksByPid.set(pid, currentLocalTracks);
    for (let trackIndex = 0; trackIndex < localTracks.length; trackIndex++) {
      const localTrack = localTracks[trackIndex];
      const trackType = localTrack.type;

      switch (trackType) {
        case 'network':
        case 'memory':
        case 'ipc': {
          // Hide those global track types because we want to hide as much as
          // possible from web developers for now.
          currentLocalTracks.add(trackIndex);
          break;
        }
        case 'thread': {
          // We don't want to display empty tracks if the tab filtered thread is empty.
          if (
            localTrack.threadIndex !== undefined &&
            localTrack.threadIndex !== null &&
            isTabFilteredThreadEmpty(localTrack.threadIndex, state)
          ) {
            // Thread is empty and we should hide it.
            currentLocalTracks.add(trackIndex);
          }
          break;
        }
        default:
          throw assertExhaustiveCheck(trackType, `Unhandled LocalTrack type.`);
      }
    }
  }

  return activeTabHiddenLocalTracksByPid;
}

/**
 * Checks whether the tab filtered thread is empty or not.
 * We take a look at the sample and marker data to determine that.
 */
function isTabFilteredThreadEmpty(
  threadIndex: ThreadIndex,
  state: State
): boolean {
  // Have to get the thread selectors to look if the thread is empty or not.
  const threadSelectors = getThreadSelectors(threadIndex);
  const tabFilteredThread = threadSelectors.getActiveTabFilteredThread(state);
  // Check the samples first to see if they are all empty or not.
  for (const stackIndex of tabFilteredThread.samples.stack) {
    if (stackIndex !== null) {
      // Samples are not empty. Do not hide that thread.
      // We don't have to look at the markers because samples are not empty.
      return false;
    }
  }

  const tabFilteredMarkers = threadSelectors.getActiveTabFilteredMarkerIndexesWithoutGlobals(
    state
  );
  if (tabFilteredMarkers.length > 0) {
    // Thread has some markers in it. Don't hide and skip to the next global track.
    return false;
  }

  return true;
}
