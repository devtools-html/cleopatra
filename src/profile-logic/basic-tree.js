/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

export class BasicTree<DataType, DisplayData> {
  _data: DataType[];
  _displayDataByIndex: Map<number, DisplayData>;

  constructor(data: DataType[]) {
    this._data = data;
    this._displayDataByIndex = new Map();
  }

  getRoots(): number[] {
    return this._data.map((_, i) => i);
  }

  getChildren(index: number): number[] {
    return index === -1 ? this.getRoots() : [];
  }

  hasChildren(_index: number): boolean {
    return false;
  }

  getAllDescendants(): Set<number> {
    return new Set();
  }

  getParent(): number {
    // -1 isn't used, but needs to be compatible with the call tree.
    return -1;
  }

  getDepth() {
    return 0;
  }

  hasSameNodeIds(tree: BasicTree<DataType, DisplayData>) {
    return this._data === tree._data;
  }

  getDisplayData(_index: number): DisplayData {
    throw new Error('Please implement `getDisplayData` in your extension.');
  }
}
