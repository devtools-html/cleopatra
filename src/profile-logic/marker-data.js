/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type {
  SamplesTable,
  MarkersTable,
  IndexIntoStringTable,
} from '../types/profile';
import type { TracingMarker } from '../types/profile-derived';
import type { BailoutPayload, ScreenshotPayload } from '../types/markers';
import type { StartEndRange } from '../types/units';
import type { UniqueStringArray } from '../utils/unique-string-array';
import { getNumberPropertyOrNull } from '../utils/flow';

export function getJankInstances(
  samples: SamplesTable,
  thresholdInMs: number
): TracingMarker[] {
  const addTracingMarker = () =>
    jankInstances.push({
      start: lastTimestamp - lastResponsiveness,
      dur: lastResponsiveness,
      title: `${lastResponsiveness.toFixed(2)}ms event processing delay`,
      name: 'Jank',
      data: null,
    });

  let lastResponsiveness = 0;
  let lastTimestamp = 0;
  const jankInstances = [];
  for (let i = 0; i < samples.length; i++) {
    const currentResponsiveness = samples.responsiveness[i];
    if (currentResponsiveness === null || currentResponsiveness === undefined) {
      return [];
    }
    if (currentResponsiveness < lastResponsiveness) {
      if (lastResponsiveness >= thresholdInMs) {
        addTracingMarker();
      }
    }
    lastResponsiveness = currentResponsiveness;
    lastTimestamp = samples.time[i];
  }
  if (lastResponsiveness >= thresholdInMs) {
    addTracingMarker();
  }
  return jankInstances;
}

export function getSearchFilteredTracingMarkers(
  markers: TracingMarker[],
  searchString: string
): TracingMarker[] {
  if (!searchString) {
    return markers;
  }
  const lowerCaseSearchString = searchString.toLowerCase();
  const newMarkers: TracingMarker[] = [];
  for (const marker of markers) {
    const { data, name } = marker;
    const lowerCaseName = name.toLowerCase();
    if (lowerCaseName.includes(lowerCaseSearchString)) {
      newMarkers.push(marker);
      continue;
    }
    if (data && typeof data === 'object') {
      if (
        typeof data.eventType === 'string' &&
        data.eventType.toLowerCase().includes(lowerCaseSearchString)
      ) {
        // Match DOMevents data.eventType
        newMarkers.push(marker);
        continue;
      }
      if (
        typeof data.name === 'string' &&
        data.name.toLowerCase().includes(lowerCaseSearchString)
      ) {
        // Match UserTiming's name.
        newMarkers.push(marker);
        continue;
      }
      if (
        typeof data.category === 'string' &&
        data.category.toLowerCase().includes(lowerCaseSearchString)
      ) {
        // Match UserTiming's name.
        newMarkers.push(marker);
        continue;
      }
    }
  }
  return newMarkers;
}

/**
 * This function takes a marker that packs in a marker payload into the string of the
 * name. This extracts that and turns it into a payload.
 */
export function extractMarkerDataFromName(
  markers: MarkersTable,
  stringTable: UniqueStringArray
): MarkersTable {
  const newMarkers: MarkersTable = {
    data: markers.data.slice(),
    name: markers.name.slice(),
    time: markers.time.slice(),
    length: markers.length,
  };

  // Match: "Bailout_MonitorTypes after add on line 1013 of self-hosted:1008"
  // Match: "Bailout_TypeBarrierO at jumptarget on line 1490 of resource://devtools/shared/base-loader.js -> resource://devtools/client/shared/vendor/immutable.js:1484"
  const bailoutRegex =
    // Capture groups:
    //       type   afterAt    where        bailoutLine  script functionLine
    //        ↓     ↓          ↓                  ↓        ↓    ↓
    /^Bailout_(\w+) (after|at) ([\w _-]+) on line (\d+) of (.*):(\d+)$/;

  // Match: "Invalidate resource://devtools/shared/base-loader.js -> resource://devtools/client/shared/vendor/immutable.js:3662"
  // Match: "Invalidate self-hosted:4032"
  const invalidateRegex =
    // Capture groups:
    //         url    line
    //           ↓    ↓
    /^Invalidate (.*):(\d+)$/;

  const bailoutStringIndex = stringTable.indexForString('Bailout');
  const invalidationStringIndex = stringTable.indexForString('Invalidate');
  for (let markerIndex = 0; markerIndex < markers.length; markerIndex++) {
    const nameIndex = markers.name[markerIndex];
    const time = markers.time[markerIndex];
    const name = stringTable.getString(nameIndex);
    let matchFound = false;
    if (name.startsWith('Bailout_')) {
      matchFound = true;
      const match = name.match(bailoutRegex);
      if (!match) {
        console.error(`Could not match regex for bailout: "${name}"`);
      } else {
        const [
          ,
          type,
          afterAt,
          where,
          bailoutLine,
          script,
          functionLine,
        ] = match;
        newMarkers.name[markerIndex] = bailoutStringIndex;
        newMarkers.data[markerIndex] = ({
          type: 'Bailout',
          bailoutType: type,
          where: afterAt + ' ' + where,
          script: script,
          bailoutLine: +bailoutLine,
          functionLine: +functionLine,
          startTime: time,
          endTime: time,
        }: BailoutPayload);
      }
    } else if (name.startsWith('Invalidate ')) {
      matchFound = true;
      const match = name.match(invalidateRegex);
      if (!match) {
        console.error(`Could not match regex for bailout: "${name}"`);
      } else {
        const [, url, line] = match;
        newMarkers.name[markerIndex] = invalidationStringIndex;
        newMarkers.data[markerIndex] = {
          type: 'Invalidation',
          url,
          line,
          startTime: time,
          endTime: time,
        };
      }
    }
    if (matchFound && markers.data[markerIndex]) {
      console.error(
        "A marker's payload was rewritten based off the text content of the marker. " +
          "perf.html assumed that the payload was empty, but it turns out it wasn't. " +
          'This is most likely an error and should be fixed. The marker name is:',
        name
      );
    }
  }

  return newMarkers;
}

export function getTracingMarkers(
  markers: MarkersTable,
  stringTable: UniqueStringArray
): TracingMarker[] {
  const tracingMarkers: TracingMarker[] = [];
  // This map is used to track start and end markers for tracing markers.
  const openMarkers: Map<IndexIntoStringTable, TracingMarker[]> = new Map();
  for (let i = 0; i < markers.length; i++) {
    const data = markers.data[i];
    if (!data) {
      // Add a marker with a zero duration
      const marker = {
        start: markers.time[i],
        dur: 0,
        name: stringTable.getString(markers.name[i]),
        title: null,
        data: null,
      };
      tracingMarkers.push(marker);
    } else if (data.type === 'tracing' && data.interval) {
      // Tracing markers are created from two distinct markers that are created at
      // the start and end of whatever code that is running that we care about.
      // This is implemented by AutoProfilerTracing in Gecko.
      //
      // In this function we convert both of these raw markers into a single
      // tracing marker with a non-null duration.
      //
      // We also handle nested markers by assuming markers of the same type are
      // never interwoven: given input markers startA, startB, endC, endD, we'll
      // get 2 markers A-D and B-C.

      const time = markers.time[i];
      const nameStringIndex = markers.name[i];
      if (data.interval === 'start') {
        let markerBucket = openMarkers.get(nameStringIndex);
        if (markerBucket === undefined) {
          markerBucket = [];
          openMarkers.set(nameStringIndex, markerBucket);
        }

        markerBucket.push({
          start: time,
          name: stringTable.getString(nameStringIndex),
          dur: 0,
          title: null,
          data,
        });
      } else if (data.interval === 'end') {
        const markerBucket = openMarkers.get(nameStringIndex);
        let marker;
        if (markerBucket && markerBucket.length) {
          // We already encountered a matching "start" marker for this "end".
          marker = markerBucket.pop();
        } else {
          // No matching "start" marker has been encountered before this "end",
          // this means it was issued before the capture started. Here we create
          // a fake "start" marker to create the final tracing marker.
          // Note we won't have additional data (eg the cause stack) for this
          // marker because that data is contained in the "start" marker.

          const nameStringIndex = markers.name[i];

          marker = {
            start: -1, // Something negative so that we can distinguish it later
            name: stringTable.getString(nameStringIndex),
            dur: 0,
            title: null,
            data,
          };
        }
        if (marker.start !== undefined) {
          marker.dur = time - marker.start;
        }
        tracingMarkers.push(marker);
      }
    } else {
      // `data` here is a union of different shaped objects, that may or not have
      // certain properties. Flow doesn't like us arbitrarily accessing properties
      // that may not exist, so use a utility function to generically get the data out.
      const startTime = getNumberPropertyOrNull(data, 'startTime');
      const endTime = getNumberPropertyOrNull(data, 'endTime');

      if (startTime !== null && endTime !== null) {
        // Construct a tracing marker with a duration if these properties exist.
        const name = stringTable.getString(markers.name[i]);
        const duration = endTime - startTime;
        tracingMarkers.push({
          start: startTime,
          dur: duration,
          name,
          data,
          title: null,
        });
      } else {
        // Ensure all markers are converted to tracing markers, even if they have no
        // more timing information. This ensures that markers can be filtered by time
        // in a consistent manner.
        tracingMarkers.push({
          start: markers.time[i],
          dur: 0,
          name: stringTable.getString(markers.name[i]),
          data,
          title: null,
        });
      }
    }
  }

  // Loop over tracing "start" markers without any "end" markers
  for (const markerBucket of openMarkers.values()) {
    for (const marker of markerBucket) {
      marker.dur = Infinity;
      tracingMarkers.push(marker);
    }
  }

  tracingMarkers.sort((a, b) => a.start - b.start);
  return tracingMarkers;
}

export function filterTracingMarkersToRange(
  tracingMarkers: TracingMarker[],
  rangeStart: number,
  rangeEnd: number
): TracingMarker[] {
  return tracingMarkers.filter(
    tm => tm.start < rangeEnd && tm.start + tm.dur >= rangeStart
  );
}

export function isNetworkMarker(marker: TracingMarker): boolean {
  return !!(marker.data && marker.data.type === 'Network');
}

export function filterForNetworkChart(markers: TracingMarker[]) {
  return markers.filter(marker => isNetworkMarker(marker));
}

export function filterForMarkerChart(markers: TracingMarker[]) {
  return markers.filter(marker => !isNetworkMarker(marker));
}
// Firefox emits separate start and end markers for each load. It does this so that,
// if a profile is collected while a request is in progress, the profile will still contain
// a start marker for that request. So by looking at start markers we can get
// information about requests that were in progress at profile collection time.
// For requests that have finished, we want to merge the request's start and end
// markers into one marker.
export function mergeStartAndEndNetworkMarker(
  markers: TracingMarker[]
): TracingMarker[] {
  const sortedMarkers: TracingMarker[] = markers.slice(0);
  const filteredMarkers: TracingMarker[] = [];

  // Sort markers, alphabetized by name to filter for markers with the same name
  sortedMarkers.sort((a, b) => {
    if (a.name < b.name) return -1;
    if (a.name > b.name) return 1;
    return 0;
  });
  for (let i = 0; i < sortedMarkers.length; i++) {
    const marker = sortedMarkers[i];
    const markerNext = sortedMarkers[i + 1];

    if (!marker.data || marker.data.type !== 'Network') {
      continue;
    }
    // The timestamps on the start and end markers describe two non-overlapping parts
    // of the same load. The start marker has a duration from channel-creation until Start
    // (i.e. AsyncOpen()). The End marker has a duration from AsyncOpen time until
    // OnStopRequest.
    // In the merged marker, we want to represent the entire duration, from channel-creation
    // until OnStopRequest.
    if (markerNext !== undefined && marker.name === markerNext.name) {
      if (
        marker.data &&
        marker.data.type === 'Network' &&
        markerNext.data &&
        markerNext.data.type === 'Network'
      ) {
        // Markers have either status STATUS_START or (STATUS_STOP || STATUS_REDIRECT || STATUS_READ). STATUS_START is reliable and the only not likely to change.
        if (
          (marker.data.status === 'STATUS_START' &&
            markerNext.data.status !== 'STATUS_START') ||
          (markerNext.data.status === 'STATUS_START' &&
            marker.data.status !== 'STATUS_START')
        ) {
          // As we discard the start marker, but want the whole duration we override the
          // start of the end marker with the start time of the start marker
          const [startMarker, endMarker] =
            marker.data.status === 'STATUS_START'
              ? [marker, markerNext]
              : [markerNext, marker];
          const mergedMarker = {
            data: endMarker.data,
            dur: endMarker.dur,
            name: endMarker.name,
            title: endMarker.title,
            start: startMarker.start,
          };
          filteredMarkers.push(mergedMarker);
          i++;
        }
        continue;
      }
    }
    filteredMarkers.push(marker);
  }
  // Sort markers by startTime to display in the right order in the network panel waterfall
  filteredMarkers.sort((a, b) => a.start - b.start);
  return filteredMarkers;
}

export function extractScreenshotsById(
  markers: MarkersTable,
  stringTable: UniqueStringArray,
  rootRange: StartEndRange
): Map<string, TracingMarker[]> {
  const idToScreenshotMarkers = new Map();
  const name = 'CompositorScreenshot';
  const nameIndex = stringTable.indexForString(name);
  for (let markerIndex = 0; markerIndex < markers.length; markerIndex++) {
    if (markers.name[markerIndex] === nameIndex) {
      // Coerce the payload to a screenshot one. Don't do a runtime check that
      // this is correct.
      const data: ScreenshotPayload = (markers.data[markerIndex]: any);

      let tracingMarkers = idToScreenshotMarkers.get(data.windowID);
      if (tracingMarkers === undefined) {
        tracingMarkers = [];
        idToScreenshotMarkers.set(data.windowID, tracingMarkers);
      }

      tracingMarkers.push({
        start: markers.time[markerIndex],
        dur: 0,
        title: null,
        name,
        data,
      });

      if (tracingMarkers.length > 1) {
        // Set the duration
        const prevMarker = tracingMarkers[tracingMarkers.length - 2];
        const nextMarker = tracingMarkers[tracingMarkers.length - 1];
        prevMarker.dur = nextMarker.start - prevMarker.start;
      }
    }
  }

  for (const [, tracingMarkers] of idToScreenshotMarkers) {
    // This last marker must exist.
    const lastMarker = tracingMarkers[tracingMarkers.length - 1];
    lastMarker.dur = rootRange.end - lastMarker.start;
  }

  return idToScreenshotMarkers;
}
