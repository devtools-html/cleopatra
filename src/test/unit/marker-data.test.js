/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { getTracingMarkers } from '../../profile-logic/marker-data';
import { processProfile } from '../../profile-logic/process-profile';
import getGeckoProfile from '.././fixtures/profiles/gecko-profile';

describe('getTracingMarkers', function() {
  const profile = processProfile(getGeckoProfile());
  const thread = profile.threads[0]; // This is the parent process main thread
  const contentThread = profile.threads[2]; // This is the content process main thread
  const tracingMarkers = getTracingMarkers(
    thread.markers,
    thread.stringTable,
    thread.samples.time[0]
  );
  const contentTracingMarkers = getTracingMarkers(
    contentThread.markers,
    contentThread.stringTable,
    thread.samples.time[0]
  );

  it('creates a reasonable processed profile', function() {
    expect(thread.name).toBe('GeckoMain');
    expect(thread.processType).toBe('default');
    expect(contentThread.name).toBe('GeckoMain');
    expect(contentThread.processType).toBe('tab');
  });
  it('creates 12 tracing markers given the test data', function() {
    expect(tracingMarkers.length).toEqual(12);
  });
  it('creates a tracing marker even if there is no start or end time', function() {
    expect(tracingMarkers[1]).toMatchObject({
      start: 2,
      dur: 0,
      name: 'VsyncTimestamp',
      title: null,
    });
  });
  it('should create a tracing marker', function() {
    expect(tracingMarkers[2]).toMatchObject({
      start: 3,
      dur: 5,
      name: 'Reflow',
      title: null,
    });
  });
  it('should fold the two reflow markers into one tracing marker', function() {
    expect(tracingMarkers.length).toEqual(12);
    expect(tracingMarkers[2]).toMatchObject({
      start: 3,
      dur: 5,
      name: 'Reflow',
      title: null,
    });
  });
  it('should fold the two Rasterize markers into one tracing marker, after the reflow tracing marker', function() {
    expect(tracingMarkers[3]).toMatchObject({
      start: 4,
      dur: 1,
      name: 'Rasterize',
      title: null,
    });
  });
  it('should create a tracing marker for the MinorGC startTime/endTime marker', function() {
    expect(tracingMarkers[5]).toMatchObject({
      start: 11,
      dur: 1,
      name: 'MinorGC',
      title: null,
    });
  });
  it('should create a tracing marker for the DOMEvent marker', function() {
    expect(tracingMarkers[4]).toMatchObject({
      dur: 1,
      name: 'DOMEvent',
      start: 9,
      title: null,
    });
  });
  it('should create a tracing marker for the marker UserTiming', function() {
    expect(tracingMarkers[6]).toMatchObject({
      dur: 1,
      name: 'UserTiming',
      start: 12,
      title: null,
    });
  });
  it('should handle tracing markers without a start', function() {
    expect(tracingMarkers[0]).toMatchObject({
      start: 0, // Truncated to the time of the first captured sample.
      dur: 1,
      name: 'Rasterize',
      title: null,
    });
  });
  it('should handle tracing markers without an end', function() {
    expect(tracingMarkers[9]).toMatchObject({
      start: 20,
      dur: Infinity,
      name: 'Rasterize',
      title: null,
    });
  });
  it('should handle nested tracing markers correctly', function() {
    expect(tracingMarkers[7]).toMatchObject({
      start: 13,
      dur: 5,
      name: 'Reflow',
      title: null,
    });
    expect(tracingMarkers[8]).toMatchObject({
      start: 14,
      dur: 1,
      name: 'Reflow',
      title: null,
    });
  });
  it('should handle arbitrary event tracing markers correctly', function() {
    expect(tracingMarkers[10]).toMatchObject({
      start: 21,
      dur: 0,
      name: 'ArbitraryName',
      title: null,
      data: { category: 'ArbitraryCategory', type: 'tracing' },
    });
  });
  it('shifts content process marker times correctly', function() {
    expect(thread.processStartupTime).toBe(0);
    expect(contentThread.processStartupTime).toBe(1000);
    expect(tracingMarkers[11]).toEqual({
      data: {
        type: 'Network',
        startTime: 22,
        endTime: 24,
        id: 388634410746504,
        status: 'STATUS_STOP',
        pri: -20,
        count: 37838,
        URI: 'https://github.com/rustwasm/wasm-bindgen/issues/2',
        domainLookupStart: 22.1,
        domainLookupEnd: 22.2,
        connectStart: 22.3,
        tcpConnectEnd: 22.4,
        secureConnectionStart: 22.5,
        connectEnd: 22.6,
        requestStart: 22.7,
        responseStart: 22.8,
        responseEnd: 22.9,
      },
      dur: 2,
      name: 'Load 32: https://github.com/rustwasm/wasm-bindgen/issues/5',
      start: 22,
      title: null,
    });
    expect(contentTracingMarkers[11]).toEqual({
      data: {
        type: 'Network',
        startTime: 1022,
        endTime: 1024,
        id: 388634410746504,
        status: 'STATUS_STOP',
        pri: -20,
        count: 37838,
        URI: 'https://github.com/rustwasm/wasm-bindgen/issues/2',
        domainLookupStart: 1022.1,
        domainLookupEnd: 1022.2,
        connectStart: 1022.3,
        tcpConnectEnd: 1022.4,
        secureConnectionStart: 1022.5,
        connectEnd: 1022.6,
        requestStart: 1022.7,
        responseStart: 1022.8,
        responseEnd: 1022.9,
      },
      dur: 2,
      name: 'Load 32: https://github.com/rustwasm/wasm-bindgen/issues/5',
      start: 1022,
      title: null,
    });
  });
});
