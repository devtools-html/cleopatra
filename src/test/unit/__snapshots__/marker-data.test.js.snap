// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deriveMarkersFromRawMarkerTable matches the snapshot 1`] = `
Array [
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "interval": "end",
      "type": "tracing",
    },
    "end": 1,
    "incomplete": true,
    "name": "Rasterize",
    "start": 0,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "type": "VsyncTimestamp",
    },
    "end": null,
    "name": "VsyncTimestamp",
    "start": 2,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "cause": Object {
        "stack": 2,
        "time": 1,
      },
      "interval": "start",
      "type": "tracing",
    },
    "end": 8,
    "name": "Reflow",
    "start": 3,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "interval": "start",
      "type": "tracing",
    },
    "end": 5,
    "name": "Rasterize",
    "start": 4,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "DOMEvent",
      "eventType": "mouseout",
      "interval": "start",
      "phase": 3,
      "timeStamp": 1,
      "type": "tracing",
    },
    "end": 10,
    "name": "DOMEvent",
    "start": 9,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "type": "GCMinor",
    },
    "end": 12,
    "name": "MinorGC",
    "start": 11,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "entryType": "measure",
      "name": "processing-thread",
      "type": "UserTiming",
    },
    "end": 13,
    "name": "UserTiming",
    "start": 12,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "cause": Object {
        "stack": 2,
        "time": 1,
      },
      "interval": "start",
      "type": "tracing",
    },
    "end": 18,
    "name": "Reflow",
    "start": 13,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "cause": Object {
        "stack": 2,
        "time": 1,
      },
      "interval": "start",
      "type": "tracing",
    },
    "end": 15,
    "name": "Reflow",
    "start": 14,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "ArbitraryCategory",
      "type": "tracing",
    },
    "end": null,
    "name": "ArbitraryName",
    "start": 21,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "URI": "https://github.com/rustwasm/wasm-bindgen/issues/5",
      "cause": undefined,
      "connectEnd": 23.6,
      "connectStart": 23.3,
      "count": 37838,
      "domainLookupEnd": 23.2,
      "domainLookupStart": 23.1,
      "endTime": 24,
      "fetchStart": 23,
      "id": 388634410746504,
      "pri": -20,
      "requestStart": 23.7,
      "responseEnd": 23.9,
      "responseStart": 23.8,
      "secureConnectionStart": 23.5,
      "startTime": 22,
      "status": "STATUS_STOP",
      "tcpConnectEnd": 23.4,
      "type": "Network",
    },
    "end": 24,
    "name": "Load 32: https://github.com/rustwasm/wasm-bindgen/issues/5",
    "start": 22,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "cause": Object {
        "stack": 2,
        "time": 1,
      },
      "filename": "/foo/bar/",
      "operation": "create/open",
      "source": "PoisionOIInterposer",
      "type": "FileIO",
    },
    "end": 24,
    "name": "FileIO",
    "start": 22,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "type": "CompositorScreenshot",
      "url": 16,
      "windowHeight": 1000,
      "windowID": "0x136888400",
      "windowWidth": 1280,
    },
    "end": 25,
    "name": "CompositorScreenshot",
    "start": 25,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "prefAccessTime": 114.9,
      "prefKind": "User",
      "prefName": "layout.css.dpi",
      "prefType": "Int",
      "prefValue": "-1",
      "type": "PreferenceRead",
    },
    "end": 27,
    "name": "PreferenceRead",
    "start": 26,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "direction": "sending",
      "endTime": 1031,
      "messageSeqno": 1,
      "messageType": "PContent::Msg_PreferenceUpdate",
      "otherPid": 2222,
      "phase": "endpoint",
      "recvEndTime": 1030.3,
      "recvThreadName": "Content Process (Thread ID: 1111)",
      "recvTid": 1111,
      "sendEndTime": 30.2,
      "sendStartTime": 30.1,
      "sendThreadName": "Parent Process (Thread ID: 3333)",
      "sendTid": 3333,
      "side": "parent",
      "startTime": 30,
      "sync": false,
      "type": "IPC",
    },
    "end": 1031,
    "incomplete": false,
    "name": "IPCOut",
    "start": 30,
    "title": "IPC — sent to Content Process (Thread ID: 1111)",
  },
  Object {
    "category": 0,
    "data": Object {
      "direction": "sending",
      "endTime": 1031,
      "messageSeqno": 1,
      "messageType": "PContent::Msg_PreferenceUpdate",
      "otherPid": 2222,
      "phase": "transferStart",
      "recvEndTime": 1030.3,
      "recvThreadName": "Content Process (Thread ID: 1111)",
      "recvTid": 1111,
      "sendEndTime": 30.2,
      "sendStartTime": 30.1,
      "sendThreadName": "Parent Process (Thread ID: 3333)",
      "sendTid": 3333,
      "side": "parent",
      "startTime": 30,
      "sync": false,
      "type": "IPC",
    },
    "end": 1031,
    "incomplete": false,
    "name": "IPCOut",
    "start": 30,
    "title": "IPC — sent to Content Process (Thread ID: 1111)",
  },
  Object {
    "category": 0,
    "data": Object {
      "direction": "sending",
      "endTime": undefined,
      "messageSeqno": 2,
      "messageType": "PContent::Msg_PreferenceUpdate",
      "otherPid": 9999,
      "phase": "endpoint",
      "recvEndTime": undefined,
      "recvThreadName": undefined,
      "recvTid": undefined,
      "sendEndTime": undefined,
      "sendStartTime": undefined,
      "sendThreadName": "Parent Process (Thread ID: 3333)",
      "sendTid": 3333,
      "side": "parent",
      "startTime": 40,
      "sync": false,
      "type": "IPC",
    },
    "end": 40,
    "incomplete": true,
    "name": "IPCOut",
    "start": 40,
    "title": "IPC — sent to 9999",
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Paint",
      "interval": "start",
      "type": "tracing",
    },
    "end": 100,
    "incomplete": true,
    "name": "Rasterize",
    "start": 100,
    "title": null,
  },
]
`;
